src/strategy/marker_file.rs

```diff
@@ -0,0 +1,227 @@
+//! Strategy for local marker file-based updates.
+
+use anyhow::{Context, Error, Result};
+use fn_error_context::context;
+use futures::future;
+use futures::prelude::*;
+use log::trace;
+use serde::{Deserialize, Serialize};
+#[cfg(not(test))]
+use std::os::unix::fs::MetadataExt;
+use std::os::unix::fs::PermissionsExt;
+use std::pin::Pin;
+use std::time::{SystemTime, UNIX_EPOCH};
+use tokio::fs as tokio_fs;
+
+/// Struct to parse finalization marker file's JSON content into.
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct FinalizationMarker {
+    /// Unix timestamp of expiry time.
+    allow_until: Option<u64>,
+}
+
+/// Strategy for immediate updates.
+#[derive(Clone, Debug, Default, Serialize)]
+pub(crate) struct StrategyMarkerFile {}
+
+impl StrategyMarkerFile {
+    /// Strategy label/name.
+    pub const LABEL: &'static str = "marker_file";
+    /// Local filesystem path to finalization marker file.
+    pub const FINALIZATION_MARKER_FILE_PATH: &'static str =
+        "/var/lib/zincati/admin/marker_file_strategy/allowfinalize.json";
+
+    /// Check if finalization is allowed.
+    pub(crate) fn can_finalize(&self) -> Pin<Box<dyn Future<Output = Result<bool, Error>>>> {
+        Box::pin(Self::marker_file_allow_finalization(
+            Self::FINALIZATION_MARKER_FILE_PATH,
+        ))
+    }
+
+    /// Try to report steady state.
+    pub(crate) fn report_steady(&self) -> Pin<Box<dyn Future<Output = Result<bool, Error>>>> {
+        trace!("marker_file strategy, report steady: {}", true);
+
+        let res = future::ok(true);
+        Box::pin(res)
+    }
+
+    /// Asynchronous helper function that returns a future indicating whether
+    /// finalization is allowed, depending on the presence of a marker file.
+    async fn marker_file_allow_finalization(
+        finalization_marker_path: &'static str,
+    ) -> Result<bool> {
+        if !verify_file_metadata(finalization_marker_path).await? {
+            return Ok(false);
+        }
+
+        if is_expired(finalization_marker_path).await? {
+            return Ok(false);
+        }
+
+        Ok(true)
+    }
+}
+
+/// Verify that finalization marker file exists, is a regular file,
+/// has the correct permissions, and is owned by the root user.
+#[context("failed to verify finalization marker file metadata")]
+async fn verify_file_metadata(path: &str) -> Result<bool> {
+    let attr = tokio_fs::metadata(path).await;
+    let attr = match attr {
+        Ok(attr) => attr,
+        // If `path` doesn't exist, return false early.
+        Err(_) => return Ok(false),
+    };
+
+    if !attr.is_file() {
+        anyhow::bail!("file is not regular file");
+    }
+
+    ensure_owned_by_root(&attr)?;
```

Comment:
Do we still need this check if `/var/lib/zincati` is owned by the `zincati` user already anyway?

Reply:
No, we trust systemd-tmpfiles for that (also, it's probably fine if the hierarchy is owned by `root` either).